schema-registry/src/main/java/io/streamnative/pulsar/handlers/kop/schemaregistry/model/CompatibilityChecker.java

```diff
@@ -0,0 +1,185 @@
+import io.apicurio.registry.rules.compatibility.JsonSchemaCompatibilityChecker;
+import io.apicurio.registry.rules.compatibility.NoopCompatibilityChecker;
+import io.apicurio.registry.rules.compatibility.ProtobufCompatibilityChecker;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+import lombok.experimental.UtilityClass;
+import lombok.extern.slf4j.Slf4j;
+
+@UtilityClass
+@Slf4j
+public class CompatibilityChecker {
+
+    /**
+     * Verify the compatibility of a Schema, following the request mode.
+     * @param schema
+     * @param subject
+     * @param schemaStorage
+     * @return
+     */
+    public static CompletableFuture<Boolean> verify(Schema schema, String subject, SchemaStorage schemaStorage) {
+        log.info("verify {} {}", subject, schema.getSchemaDefinition());
+        CompletableFuture<CompatibilityChecker.Mode> mode = schemaStorage.getCompatibilityMode(subject);
+        return mode.thenCompose(m -> {
+            return verifyCompatibility(schema, subject, schemaStorage, m);
+        });
+    }
+
+    private static CompletableFuture<Boolean> verifyCompatibility(Schema schema, String subject,
+                                                                  SchemaStorage schemaStorage, Mode mode) {
+        log.info("verify {} {} mode ", subject, mode);
+        if (mode == Mode.NONE) {
+            return CompletableFuture.completedFuture(true);
+        }
+        CompletableFuture<List<Integer>> versions = schemaStorage.getAllVersionsForSubject(subject);
+        return versions.thenCompose(vv -> {
+            return verifyCompatibility(schema, schemaStorage, mode, vv);
+        });
+    }
+
+    private static CompletableFuture<Boolean> verifyCompatibility(Schema schema, SchemaStorage schemaStorage,
+                                                                  Mode mode, List<Integer> versions) {
+        if (versions.isEmpty()) {
+            // no versions ?
+            return CompletableFuture.completedFuture(true);
+        }
+        final List<Integer> idsToCheck;
+        if (mode == Mode.BACKWARD || mode == Mode.FORWARD) {
+            // only latest
+            idsToCheck = Collections.singletonList(versions.stream().mapToInt(Integer::intValue).max().getAsInt());
+        } else {
+            // all the versions
+            idsToCheck = versions;
+        }
+        log.info("Compare schema against {} ids", idsToCheck);
+
+        CompletableFuture<List<Schema>>
+                res = schemaStorage.downloadSchemas(idsToCheck);
+
+        return res.thenApply((downloadedSchemas) -> {
+            return verify(schema.getSchemaDefinition(), schema.getType(), mode, downloadedSchemas);
+        });
+    }
+
+    public static boolean verify(String schemaDefinition, String type, Mode mode, List<Schema> allSchemas) {
```

Comment:
Codacy reported this method has an NPath complexity of 224 (current threshold is 200).

We can add a method into `Mode`:

```java
        public CompatibilityLevel toCompatibilityLevel() {
            switch (this) {
                case BACKWARD:
                    return CompatibilityLevel.BACKWARD;
                case BACKWARD_TRANSITIVE:
                    return CompatibilityLevel.BACKWARD_TRANSITIVE;
                case FORWARD:
                    return CompatibilityLevel.FORWARD;
                case FORWARD_TRANSITIVE:
                    return CompatibilityLevel.FORWARD_TRANSITIVE;
                case FULL:
                    return CompatibilityLevel.FULL;
                case FULL_TRANSITIVE:
                    return CompatibilityLevel.FULL_TRANSITIVE;
                default:
                    return CompatibilityLevel.NONE;
            }
        }

        public boolean checkOnlyLatest() {
            return this == BACKWARD || this == FORWARD || this == FULL || this == NONE;
        }
```

Then change the code here to:

```java
        final boolean onlyLatest = mode.checkOnlyLatest();
        final CompatibilityLevel level = mode.toCompatibilityLevel();
```

Reply:
thank you! done